accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,126 @@
+
+Source generators should not use public API with "reserved" or "undocumented" functionality in attempts to avoid public API garuntees.  Libraries paired with source generators need to provide the same garuntees as libraries alone do today.
+
+Source generators should avoid emitting public API in user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name colisions with user code and provide affordances for resolving name conflicts. 
+
+### Functionality concerns for source generators
+
+Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.
+
+Source generators target .NETStandard and must run on both .NETFramework and .NETCore.  
+
+Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.
+
+Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence garunteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.
```

Comment:
Should we also have chapter on Security concerns for source generators?

The security bugs in code generated by source generators cannot by serviced by traditional runtime servicing channel. They require rebuilding with the serviced toolchain. 

- The exact version of the source generator used to build the assembly needs to be recorded somewhere (where?), so that it is possible to flag the assembly as being built with vulnerable version of the source generator.

- The code generated by the source generator should aim to have low security risk. For example, the source generators should not be generating unsafe buffer manipulation operations inline. Instead, the source generators should prefer to call APIs in the framework to perform unsafe buffer manipulations where possible.

Reply:
Initially this doc was only talking about distribution but we might as well grow it to discuss this as well.  It's not a bad idea to preserve some metadata about the generator for identification purposes.  Should we add assembly metadata?  We can't offer any automatic updates of course.

Reply:
Assembly metadata and/or record the source generator details in PDBs are the two options that I can think of.

Reply:
I have a few more, will update the doc.